File: jugglerhelper/src/main/java/net/maxsmr/jugglerhelper/fragments/toolbar/BaseJugglerToolbarFragment.java
package net.maxsmr.jugglerhelper.fragments.toolbar;

import android.app.Activity;
import android.graphics.drawable.Drawable;
import android.os.Bundle;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import androidx.annotation.CallSuper;
import androidx.annotation.LayoutRes;
import androidx.annotation.StringRes;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.maxsmr.commonutils.android.gui.GuiUtils;
import net.maxsmr.jugglerhelper.R;
import net.maxsmr.jugglerhelper.fragments.BaseJugglerFragment;
import net.maxsmr.jugglerhelper.navigation.NavigationMode;

import java.util.List;

import me.ilich.juggler.gui.JugglerActivity;
import me.ilich.juggler.gui.JugglerToolbarFragment;
import me.ilich.juggler.states.State;

public abstract class BaseJugglerToolbarFragment extends JugglerToolbarFragment {

    @LayoutRes
    protected abstract int getLayoutId();

    @NotNull
    @Override
    public final View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(getLayoutId(), container, false);
        onBindViews(rootView);
        return rootView;
    }

    protected void onBindViews(@NotNull View rootView) {
    }

    protected void initToolbar() {
        View rootView = getView();

        if (rootView == null) {
            throw new IllegalStateException("root view was not created");
        }

        Toolbar toolbar = rootView.findViewById(getToolbarId());

        if (toolbar == null) {
            throw new IllegalStateException("toolbar was not found");
        }

        toolbar.setNavigationOnClickListener(v -> getJugglerActivity().onSupportNavigateUp());

        ActionBar actionBar = getJugglerActivity().getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle("");
            actionBar.setDisplayShowTitleEnabled(true);
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.show();
        }
    }

    protected void initTitle() {
        setTitle(getTitle());
    }

    // setting logo by layout is not working properly
    protected void initLogo() {
        setLogo(getLogo());
    }

    protected void initNavigationMode() {
        setMode(getNavigationMode(), getUpIcon());
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initToolbar();
        initTitle();
        initLogo();
        initNavigationMode();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        ActionBar actionBar = getJugglerActivity().getSupportActionBar();
        if (actionBar != null) {
            actionBar.hide();
        }
    }

    private void setMode(NavigationMode mode, @Nullable Drawable upIcon) {

        if (mode != null) {

            Toolbar toolbar = getToolbar();

            if (toolbar != null) {

                if (upIcon != null) {
                    toolbar.setNavigationIcon(upIcon);
                }

                switch (mode) {
                    case SANDWICH:
                        setNavigationIconVisible(true);
                        break;
                    case BACK:
                        setNavigationIconVisible(true);
                        break;
                    case NONE:
                        setNavigationIconVisible(false);
                        toolbar.setVisibility(View.VISIBLE);
                        break;
                    case INVISIBLE:
                        setNavigationIconVisible(false);
                        toolbar.setVisibility(View.GONE);
                        break;
                }
            }
        }
    }

    @Nullable
    protected Drawable getUpIcon() {
        State state = getState();
        return state != null ? state.getUpNavigationIcon(getContext()) : null;
    }

    private void setNavigationIconVisible(boolean b) {
        ActionBar actionBar = getJugglerActivity().getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(b);
            actionBar.setHomeButtonEnabled(b);
        }
    }

    @Nullable
    public String getTitle() {
        String title = null;
        @SuppressWarnings("unchecked") State<State.Params> state = (State<State.Params>) getState();
        if (state != null) {
            title = state.getTitle(getContext(), state.getParams());
        } else {
            Activity activity = getActivity();
            if (activity != null) {
                CharSequence activityTitle = activity.getTitle();
                title = activityTitle != null ? activityTitle.toString() : null;
            }
        }
        return title;
    }

    public void setTitle(CharSequence text) {
        Toolbar toolbar = getToolbar();
        if (toolbar != null) {
            toolbar.setTitle(text);
        }
        ActionBar actionBar = getJugglerActivity().getSupportActionBar();
        if (actionBar != null) {
            actionBar.setTitle(text);
        }
    }

    public void setTitle(@StringRes int textRestId) {
        setTitle(getString(textRestId));
    }

    @Nullable
    protected abstract Drawable getLogo();

    public void setLogo(Drawable icon) {
        Toolbar toolbar = getToolbar();
        if (toolbar == null) {
            throw new RuntimeException("toolbar was not initialized");
        }
        toolbar.setLogo(icon);
    }

    @NotNull
    protected abstract NavigationMode getNavigationMode();

    @CallSuper
    public void onStateActivated(JugglerActivity activity, State<?> state) {
        initTitle();
        FragmentManager fm = getChildFragmentManager();
        List<Fragment> fragments = fm.getFragments();
        for (Fragment fragment : fragments) {
            if (fragment instanceof BaseJugglerFragment) {
                ((BaseJugglerFragment) fragment).onStateActivated(activity, state);
            } else if (fragment instanceof BaseJugglerToolbarFragment) {
                ((BaseJugglerToolbarFragment) fragment).onStateActivated(activity, state);
            }
        }
    }

}

File: jugglerhelper/src/main/kotlin/net/maxsmr/jugglerhelper/fragments/toolbar/BaseJugglerToolbarFragment.kt
package net.maxsmr.jugglerhelper.fragments.toolbar

import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.annotation.LayoutRes
import me.ilich.juggler.gui.JugglerActivity
import me.ilich.juggler.gui.JugglerToolbarFragment
import me.ilich.juggler.states.State
import net.maxsmr.commonutils.data.text.EMPTY_STRING
import net.maxsmr.jugglerhelper.fragments.BaseJugglerFragment
import net.maxsmr.jugglerhelper.navigation.NavigationMode

abstract class BaseJugglerToolbarFragment : JugglerToolbarFragment() {

    @get:LayoutRes
    protected abstract val layoutId: Int

    protected abstract val navigationMode: NavigationMode

    @Suppress("UNCHECKED_CAST")
    protected open val title: String?
        get() {
            val state = state as State<State.Params?>?
            return if (state != null) {
                state.getTitle(requireContext(), state.params)
            } else {
                requireActivity().title?.toString()
            }
        }

    protected open val logo: Drawable? = null

    protected open val upIcon: Drawable?
        get() {
            val state = state
            return state?.getUpNavigationIcon(requireContext())
        }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
            inflater.inflate(layoutId, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initToolbar()
        initTitle()
        initLogo()
        initNavigationMode()
        initViews(view)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        val actionBar = jugglerActivity.supportActionBar
        actionBar?.hide()
    }

    @CallSuper
    fun onStateActivated(activity: JugglerActivity?, state: State<*>?) {
        initTitle()
        for (fragment in childFragmentManager.fragments) {
            if (fragment is BaseJugglerFragment) {
                fragment.onStateActivated(activity, state)
            } else if (fragment is BaseJugglerToolbarFragment) {
                fragment.onStateActivated(activity, state)
            }
        }
    }

    protected open fun initViews(view: View) {
        // override if needed
    }

    protected open fun applyTitle(title: CharSequence?) {
        toolbar?.title = title
        jugglerActivity.supportActionBar?.title = title
    }

    protected open fun applyLogo(logo: Drawable?) {
        toolbar?.logo = logo
    }

    protected open fun initToolbar() {
        toolbar?.setNavigationOnClickListener { jugglerActivity.onSupportNavigateUp() }
        jugglerActivity.supportActionBar?.let { actionBar ->
            actionBar.title = EMPTY_STRING
            actionBar.setDisplayShowTitleEnabled(true)
            actionBar.show()
        }
    }

    protected open fun initTitle() {
        applyTitle(title)
    }

    // setting logo by layout is not working properly
    protected open fun initLogo() {
        applyLogo(logo)
    }

    protected fun initNavigationMode() {
        applyMode(navigationMode, upIcon)
    }

    private fun applyMode(mode: NavigationMode, upIcon: Drawable?) {

        fun setNavigationIconVisible(b: Boolean) {
            jugglerActivity.supportActionBar?.let {
                it.setDisplayHomeAsUpEnabled(b)
                it.setHomeButtonEnabled(b)
            }
        }

        val toolbar = toolbar
        if (toolbar != null) {
            upIcon?.let {
                toolbar.navigationIcon = upIcon
            }
            when (mode) {
                NavigationMode.SANDWICH -> setNavigationIconVisible(true)
                NavigationMode.BACK -> setNavigationIconVisible(true)
                NavigationMode.NONE -> {
                    setNavigationIconVisible(false)
                    toolbar.visibility = View.VISIBLE
                }
                NavigationMode.INVISIBLE -> {
                    setNavigationIconVisible(false)
                    toolbar.visibility = View.GONE
                }
            }
        }
    }
}
